File: algorithms/0811.subdomain-visit-count/solution.rs
impl Solution {
    pub fn subdomain_visits(cpdomains: Vec<String>) -> Vec<String> {
        use std::collections::HashMap;

        let mut map: HashMap<String, i32> = HashMap::new();

        for cpdomain in cpdomains {
            let pair = (cpdomain.split_whitespace().collect::<Vec<&str>>());
            let visits = pair[0].parse::<i32>().unwrap();
            let domains = pair[1].split(".").collect::<Vec<&str>>();

            let l = domains.len();

            for i in 0..l {
                let level = &domains[i..l].join(".");

                let total_visits = map.entry(level.to_string()).or_insert(0);
                *total_visits += visits;
            }
        }

        let mut res = Vec::with_capacity(map.len());
        for (k, v) in map {
            res.push(v.to_string() + " " + &k);
        }

        res
    }
}

File: algorithms/0811.subdomain-visit-count/solution.rb
# @param {String[]} cpdomains
# @return {String[]}
def subdomain_visits(cpdomains)
  hash = Hash.new

  cpdomains.each do |cpdomain|
    visits, domains = cpdomain.split(' ')
    domains = domains.split('.')

    for i in 0...domains.size
      level = domains[i...domains.size].join('.')
      total_visits = 0
      if hash.has_key?(level)
        total_visits = hash[level]
      end
      hash[level] = total_visits + visits.to_i
    end
  end

  hash.map { |k, v| [v, k].join(' ') }
end
